### ISSUE TITLE
Keyboard on Surface Laptop go 2 does not work.

### ISSUE THREAD

**Issue by keithga** [ipxe/ipxe#975]
https://www.microsoft.com/en-us/d/surface-laptop-go-2/8PGLPV76MJHN

After booting to iPXE, attached physical on Laptop does not work. 
However, as a work-around, the "On-Screen" keyboard for the touch screen does work, as does any attached USB keyboard. 

My assumption is that the attached keyboard is not enumerated as a standard USB device, instead requiring some kind of intermediate driver. 

Attached is a list of all "HID" PNP device ID's on the Surface Laptop go 2: 

[surfacelist.txt](https://github.com/ipxe/ipxe/files/11792945/surfacelist.txt)

**Comment by NiKiZe** [comment]
If you boot directly to the uefi shell, does the keyboard work there?
Can you boot Linux and figure out how it is connected?

Without details on how to fix this (or paying for commercial support) it is unlikely to get any iPXE based fixed. More likely there can be a firmware based fix released.

**Comment by mcb30** [comment]
Built-in laptop keyboards are typically wired in as PS/2 keyboards, since the wire protocol consumes less power than USB.  iPXE does not get involved in driving the hardware for these, but will rely solely on the UEFI firmware's `ConIn` device to read keystrokes.

As @NiKiZe suggested, try booting into the UEFI shell to see if it works there.

**Comment by NiKiZe** [comment]
Does the efi shell work? What does lsusb (in Linux) show on this device?

**Comment by keithga** [comment]
Sorry, there is NO EFI shell on surface go devices (thanks Microsoft).

**Comment by NiKiZe** [comment]
Wait, isn't the keyboard on those Bluetooth!?

Boot efi shell binary from USB?

**Comment by keithga** [comment]
No, this is a Surface *LAPTOP* Go 2, the keyboard is attached at all times. 

I was surprised, just ran on my test machine, and the attached keyboard DOES work with the UEFI shell from:
https://github.com/tianocore/edk2/blob/edk2-stable201903/ShellBinPkg/UefiShell/X64/Shell.efi

**Comment by NiKiZe** [comment]
Which iPXE binary are you using? Have you built it yourself? Does it have USB support enabled (if it does it takes over all USB from firmware, including keyboard)

**Comment by keithga** [comment]
We are most likely using the 2Pint snponly_drv_x64.efi binary which has support for USB networking drivers:
o	CDC-ECM USB Ethernet driver
o	USB RNDIS Ethernet driver
o	CDC-NCM USB Ethernet driver
I will try to re-run using the 2Pint snponly_x64.efi binary, when I get a chance.

**Comment by NiKiZe** [comment]
In that case only USB devices should be disconnected, you will need to be able to build and run debug builds of iPXE to try and figure out if this is USB related or not.

**Comment by keithga** [comment]
We were able to get a hold of a Surface Laptop Go 2 today, and ran some tests.

When booted from our "snponly_x64.efi" file, we *ARE* able to type using the build-in keyboard.
When booted from our "snponly_usb_x64.efi" file, we are NOT able to type using the build-in keyboard.
The snponly_usb_x64.efi program is built with the following extra drivers:
    --ecm--acm--ncm 

So, yes it's entirely possible that these USB Network drivers are somehow interfering with the on-board USB Keyboard. 
We will use the non-USB Network driver "snponly_x64.efi" ipxe program on the Surface Laptop Go 2 as a work around for now.

Thanks

**Comment by NiKiZe** [comment]
Full lsusb -v might be helpful here (from Linux)

**Comment by aretsan** [comment]
Also encountered the same problem here. 

ipxe.efi: keyboard works properly.
ipxe--ecm.efi: keyboard does not work.
ipxe--acm.efi: keyboard does not work.
ipxe--ncm.efi: keyboard does not work.
ipxe--ecm--ncm--acm.efi: keyboard does not work.

and by the way this is the keyboard in my laptop:
```
Bus 001 Device 002: ID 0b05:1866 ASUSTek Computer, Inc. N-KEY Device
Device Descriptor:
  bLength                18
  bDescriptorType         1
  bcdUSB               2.00
  bDeviceClass            0 [unknown]
  bDeviceSubClass         0 [unknown]
  bDeviceProtocol         0
  bMaxPacketSize0        64
  idVendor           0x0b05 ASUSTek Computer, Inc.
  idProduct          0x1866 N-KEY Device
  bcdDevice           40.22
  iManufacturer           1 ASUSTek Computer Inc.
  iProduct                2 N-KEY Device
  iSerial                 0
  bNumConfigurations      1
  Configuration Descriptor:
    bLength                 9
    bDescriptorType         2
    wTotalLength       0x0022
    bNumInterfaces          1
    bConfigurationValue     1
    iConfiguration          0
    bmAttributes         0xe0
      Self Powered
      Remote Wakeup
    MaxPower              100mA
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        0
      bAlternateSetting       0
      bNumEndpoints           1
      bInterfaceClass         3 Human Interface Device
      bInterfaceSubClass      1 Boot Interface Subclass
      bInterfaceProtocol      1 Keyboard
      iInterface              0
        HID Device Descriptor:
          bLength                 9
          bDescriptorType        33
          bcdHID               1.10
          bCountryCode            0 Not supported
          bNumDescriptors         1
          bDescriptorType        34 Report
          wDescriptorLength     331
          Report Descriptors:
            ** UNAVAILABLE **
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x81  EP 1 IN
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0040  1x 64 bytes
        bInterval               4
Device Status:     0x0001
  Self Powered
```

**Comment by Sisterjohng** [comment]
May be to late on this or you may have already found it, but could be worth checking out https://github.com/linux-surface & their repos. Sounds like from what I've seen you could solve some of your issues.

**Comment by NiKiZe** [comment]
I think these are AMI and it is the same as #1643, wrong order of things here, but closing this as duplicate since the other one has more details.

What happens is that iPXE takes over the USB bus, depending on the firmware drivers they reconnect and use that bus or not.
USB keyboard driver for AMI among others fails here, now the weird thing is that external USB keyboards do work(?) but the internal does not. But these might be different drivers, using the efi shell and listing them might show something, also full lsusb would be interesting. Would be great if that data can be added.
